src/goto-programs/restrict_function_pointers.cpp

```diff
@@ -0,0 +1,415 @@
+}
+
+function_pointer_restrictionst::restrictionst
+parse_function_pointer_restrictions_from_file(
+  const std::list<std::string> &filenames,
+  message_handlert &message_handler)
+{
+  auto merged_restrictions = function_pointer_restrictionst::restrictionst{};
+  for(auto const &filename : filenames)
+  {
+    auto const restrictions =
+      function_pointer_restrictionst::read_from_file(filename, message_handler);
+    merged_restrictions = merge_function_pointer_restrictions(
+      merged_restrictions, restrictions.restrictions);
+  }
+  return merged_restrictions;
+}
+} // namespace
+
+function_pointer_restrictionst function_pointer_restrictionst::from_options(
+  const optionst &options,
+  message_handlert &message_handler)
+{
+  auto const restriction_opts =
+    options.get_list_option(RESTRICT_FUNCTION_POINTER_OPT);
+  auto const commandline_restrictions =
+    parse_function_pointer_restrictions_from_command_line(restriction_opts);
+  auto const file_restrictions = parse_function_pointer_restrictions_from_file(
+    options.get_list_option(RESTRICT_FUNCTION_POINTER_FROM_FILE_OPT),
+    message_handler);
+  return {merge_function_pointer_restrictions(
+    file_restrictions, commandline_restrictions)};
+}
+
+function_pointer_restrictionst function_pointer_restrictionst::read_from_file(
+  const std::string &filename,
+  message_handlert &message_handler)
+{
+  auto failed = [](bool failFlag) { return failFlag; };
+  function_pointer_restrictionst::restrictionst restrictions;
+  auto inFile = std::ifstream{filename};
+  jsont json;
+  if(failed(parse_json(inFile, filename, message_handler, json)))
+  {
+    throw system_exceptiont{
+      "failed to read function pointer restrictions from " + filename};
+  }
+  if(!json.is_object())
+  {
+    throw system_exceptiont{"top level item is not an object"};
+  }
+  for(auto const &kv_pair : to_json_object(json))
+  {
+    restrictions.emplace(irep_idt{kv_pair.first}, [&] {
+      if(!kv_pair.second.is_array())
+      {
+        throw system_exceptiont{"In " + filename + ", value of " +
+                                kv_pair.first + " is not an array"};
+      }
+      auto possible_targets = std::unordered_set<irep_idt>{};
+      auto const &array = to_json_array(kv_pair.second);
+      std::transform(
+        array.begin(),
+        array.end(),
+        std::inserter(possible_targets, possible_targets.end()),
+        [&](const jsont &array_element) {
+          if(!array_element.is_string())
+          {
+            throw system_exceptiont{"In " + filename + ", value of " +
+                                    kv_pair.first +
+                                    "contains a non-string array element"};
+          }
+          return irep_idt{to_json_string(array_element).value};
+        });
+      return possible_targets;
+    }());
+  }
+  return function_pointer_restrictionst{restrictions};
+}
+
+void function_pointer_restrictionst::write_to_file(
```

Comment:
:bulb: if this returned a json object then it would be possible to unit test this

Reply:
We now have functions `to_json()` and `from_json()`, and `write_to_file()` and `read_to_file()` use those. I've also added unit tests for the former two.